standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java

```diff
@@ -678,316 +215,39 @@ private void setCleanerStart(Connection dbConn, CompactionInfo info, Long timest
-        // max(TXNS.txn_id) could be min_uncommitted_txnid.
-        String s = "SELECT MIN(\"RES\".\"ID\") AS \"ID\" FROM (" +
-          " SELECT MAX(\"TXN_ID\") + 1 AS \"ID\" FROM \"TXNS\"" +
-          (useMinHistoryLevel ? "" :
-          "   UNION" +
-          " SELECT MIN(\"WS_TXNID\") AS \"ID\" FROM \"WRITE_SET\"") +
-          "   UNION" +
-          " SELECT MIN(\"TXN_ID\") AS \"ID\" FROM \"TXNS\" WHERE \"TXN_STATE\" = " + TxnStatus.ABORTED +
-          (useMinHistoryLevel ? "" :
-          "   OR \"TXN_STATE\" = " + TxnStatus.OPEN) +
-          " ) \"RES\"";
+    long minTxnIdSeenOpen = findMinTxnIdSeenOpen();
 
-        LOG.debug("Going to execute query <{}>", s);
-        rs = stmt.executeQuery(s);
-        if (!rs.next()) {
-          throw new MetaException("Transaction tables not properly initialized, no record found in TXNS");
-        }
-        long minUncommitedTxnid = Math.min(rs.getLong(1), minTxnIdSeenOpen);
-
-        // As all txns below min_uncommitted_txnid are either committed or empty_aborted, we are allowed
-        // to cleanup the entries less than min_uncommitted_txnid from the TXN_TO_WRITE_ID table.
-        s = "DELETE FROM \"TXN_TO_WRITE_ID\" WHERE \"T2W_TXNID\" < " + minUncommitedTxnid;
-        LOG.debug("Going to execute delete <{}>", s);
-        int rc = stmt.executeUpdate(s);
-        LOG.info("Removed {} rows from TXN_TO_WRITE_ID with Txn Low-Water-Mark: {}", rc, minUncommitedTxnid);
-
-        LOG.debug("Going to commit");
-        dbConn.commit();
-      } catch (SQLException e) {
-        LOG.error("Unable to delete from TXN_TO_WRITE_ID table " + e.getMessage());
-        LOG.debug("Going to rollback");
-        rollbackDBConn(dbConn);
-        checkRetryable(e, "cleanTxnToWriteIdTable");
-        throw new MetaException(DB_FAILED_TO_CONNECT + e.getMessage());
-      } finally {
-        close(rs, stmt, dbConn);
-      }
-    } catch (RetryException e) {
-      cleanTxnToWriteIdTable();
+    // First need to find the min_uncommitted_txnid which is currently seen by any open transactions.
+    // If there are no txns which are currently open or aborted in the system, then current value of
+    // max(TXNS.txn_id) could be min_uncommitted_txnid.
+    Long minTxnId = new MinUncommittedTxnIdHandler(useMinHistoryLevel).execute(dataSourceWrapper);
+    if (minTxnId == null) {
+      throw new MetaException("Transaction tables not properly initialized, no record found in TXNS");
     }
+    long minUncommitedTxnid = Math.min(minTxnId, minTxnIdSeenOpen);
+
+    // As all txns below min_uncommitted_txnid are either committed or empty_aborted, we are allowed
+    // to clean up the entries less than min_uncommitted_txnid from the TXN_TO_WRITE_ID table.
+    String sql = "DELETE FROM \"TXN_TO_WRITE_ID\" WHERE \"T2W_TXNID\" < :txnId";
+    LOG.debug("Going to execute delete <{}>", sql);
+    int rc = dataSourceWrapper.getJdbcTemplate().update(sql, new MapSqlParameterSource("txnId", minUncommitedTxnid));
```

Comment:
can't we use ParameterizedCommand here?

Reply:
Sure, we can. for the very simple cases I did not create a separate class, but to be consistent, I can create for these too.